tensornetwork/contractors/custom_path_solvers/nconinterface.py

```diff
@@ -0,0 +1,227 @@
+    # find common indices between tensor pair
+    cont_many, A_cont, B_cont = np.intersect1d(new_connects[orders[0,i]],
+                                               new_connects[orders[1,i]],  
+                                               return_indices=True)
+    temp_connects = np.append(np.delete(new_connects[orders[0,i]], A_cont), 
+                              np.delete(new_connects[orders[1,i]], B_cont))
+    cont_order = list(np.concatenate((cont_order,cont_many),axis=0))
+    
+    # build new set of connects
+    new_connects[orders[0,i]] = temp_connects
+    del new_connects[orders[1,i]]
+    
+  return cont_order
+
+#------------------------------------------------------------------------#
+def ncon_cost_check(tensor_list_in: List[np.ndarray], 
+                    connect_list_in: List[Union[List[int], Tuple[int]]], 
+                    cont_order: Optional[Union[List[int], str]] = None):
+  """
+  Checks the computational cost of an `ncon` contraction (without actually 
+  doing the contraction). Ignore the cost contributions from partial traces 
+  (which are always sub-leading).
+  Args:
+    tensors: list of the tensors in the network.
+    connects: length-N list of lists (or tuples) specifying the network 
+      connections. The jth entry of the ith list in connects labels the edge 
+      connected to the jth index of the ith tensor. Labels should be positive 
+      integers for internal indices and negative integers for free indices.
+    cont_order: optional argument to specify the order for contracting the 
+      positive indices. Defaults to ascending order if omitted. 
+  Returns:
+    float: the cost of the network contraction, given as log10(total_FLOPS).
+  """
+  
+  total_cost = None
+  num_tensors = len(tensor_list_in)
+  tensor_dims = [np.array(np.log10(tensor_list_in[ele].shape)) 
+                 for ele in range(num_tensors)]
+  connect_list = [np.array(connect_list_in[ele]) for ele in range(num_tensors)]
+  
+  # generate contraction order if necessary
+  flat_connect = np.concatenate(connect_list)
+  if cont_order is None:
+    cont_order = np.unique(flat_connect[flat_connect > 0])
+  else:
+    cont_order = np.array(cont_order)
+  
+  # do all partial traces
+  for ele in range(num_tensors):
+    uni_inds, counts = np.unique(connect_list[ele],return_counts=True)
+    tr_inds = uni_inds[counts==2]
+    temp_connect = connect_list[ele]
+    temp_dims = tensor_dims[ele]
+    for ind in tr_inds:
+      temp_dims = temp_dims[temp_connect != ind]
+      temp_connect = temp_connect[temp_connect != ind]
+      cont_order = np.delete(cont_order, np.flatnonzero(cont_order == ind))
+      
+    tensor_dims[ele] = temp_dims
+    connect_list[ele] = temp_connect
+  
+  # do all binary contractions
+  while len(cont_order) > 0:
+    # identify tensors to be contracted
+    cont_ind = cont_order[0]
+    locs = [ele for ele in range(len(connect_list)) 
+            if sum(connect_list[ele] == cont_ind) > 0]
+  
```

Comment:
you can use 
```python 
c1 = connect_list.pop(locs[0])
c2 = connect_list.pop(locs[1])
dim1 = tensor_dims.pop(locs[0])
dim2 = tensor_dims.pop(locs[1])
```
then you don't have to delete them later on from `connect_list` and `tensor_dims`.
`del connect_list[locs[0]` will mutate the input list, should probably be avoided

Reply:
As far as I can see, the `c1 = connect_list.pop(locs[0])` also mutates the `connect_list` doesn't it? As it stands, I make a copy of the `connect_list` to avoid mutating the original input to the function, but in would be nice to avoid this if possible.
Also, although the `pop` is a good idea, I believe that one has to be more careful when using it to take account that it changes the lists. The correct version would be:
```
c2 = connect_list.pop(locs[1])
c1 = connect_list.pop(locs[0])
dim2 = tensor_dims.pop(locs[1])
dim1 = tensor_dims.pop(locs[0])
```
since `locs[1] > locs[0]`.

Reply:
yes you're right, the order needs to be changed.